File: lab1/object/Trip.cs
public class Trip
{
    public int Id { get; set; }
    public string PickupLocation { get; set; }
    public string DropOffLocation { get; set; }
    public string RideType { get; set; }
    public decimal Fare { get; private set; }
    public string Status { get; set; }
    public Rider Rider { get; set; }
    public Driver Driver { get; set; }
    public int Distance { get; set; }

    public void CalculateFare()
    {
        // Simple fare calculation based on distance and ride type
        Fare = Distance * (RideType == "Luxury" ? 2.0m : 1.0m);
        Console.WriteLine($"Fare calculated: {Fare:C}");
    }

    public void AssignDriver(Driver driver)
    {
        this.Driver = driver;
        this.Status = "Driver Assigned";
        Console.WriteLine("Driver has been assigned.");
    }

    public void CompleteTrip()
    {
        this.Status = "Completed";
        Console.WriteLine("Trip has been completed.");
    }
}

File: lab02_stratergy_pattern/src/Trip.java
class Trip {
    private String pickupLocation;
    private String dropOffLocation;
    private double distance;
    private RideType rideType;
    private PaymentMethod paymentMethod;
    private NotificationService notificationService;
    private double baseFare = 10.0;

    public Trip(String pickupLocation, String dropOffLocation, double distance, RideType rideType,
                PaymentMethod paymentMethod, NotificationService notificationService) {
        this.pickupLocation = pickupLocation;
        this.dropOffLocation = dropOffLocation;
        this.distance = distance;
        this.rideType = rideType;
        this.paymentMethod = paymentMethod;
        this.notificationService = notificationService;
    }

    public String getPickupLocation() {
        return pickupLocation;
    }

    public String getDropOffLocation() {
        return dropOffLocation;
    }

    public double getDistance() {
        return distance;
    }

    public double calculateFare() {
        return rideType.calculateFare(distance, baseFare);
    }

    public void startTrip() {
        notificationService.sendNotification("Your trip has started from " + pickupLocation);
    }

    public void completeTrip() {
        double fare = calculateFare();
        paymentMethod.processPayment(fare);
        notificationService.sendNotification("Your trip is complete. Fare: " + fare);
    }
}
